src/main/java/cr0s/warpdrive/config/structures/Schematic.java

```
@@ -1,38 +1,90 @@
+import java.util.Random;
+
+import static cr0s.warpdrive.config.WarpDriveConfig.LOGGING_WORLD_GENERATION;
+
 public class Schematic extends AbstractStructure {
 	
-	protected String filename;
+	protected HashMap<String, Integer> fileNames;
+	
 	protected Replacement[] replacements;
 	protected Insertion[] insertions;
 	
 	public Schematic(final String group, final String name) {
 		super(group, name);
 	}
 	
+	public String getRandomFileName(Random random) {
+		
+		// In loadFromXmlElement, it already checked that there must be at least 1 "schematic" xml node
+		// therefore, this should not be possible
+		assert(!fileNames.isEmpty());
+		
+		int totalWeight = 0;
+		for (int weight : fileNames.values()) {
+			totalWeight += weight;
+		}
+		int result = random.nextInt(totalWeight);
+		for (Map.Entry<String, Integer> entry : fileNames.entrySet()) {
+			result -= entry.getValue();
+			if (result <= 0) {
+				return entry.getKey();
+			}
+		}
+		return fileNames.keySet().iterator().next();
+	}
+	
+	@Override
+	public boolean generate(@Nonnull final World world, @Nonnull final Random random, @Nonnull final BlockPos blockPos) {
+		return instantiate(random).generate(world, random, blockPos);
+	}
+	
+	@Override
+	public AbstractStructureInstance instantiate(final Random random) {
+		return new SchematicInstance(this, random);
+	}
+	
 	@Override
 	public boolean loadFromXmlElement(final Element element) throws InvalidXmlException {
 		super.loadFromXmlElement(element);
 		
+		List<Element> fileNameList = XmlFileManager.getChildrenElementByTagName(element, "schematic");
+		if (fileNameList.isEmpty()) {
+			throw new InvalidXmlException("Must have one schematic node with file name!");
+		}
+		this.fileNames = new HashMap<>(fileNameList.size());
+		for (Element entry : fileNameList) {
+			String name = entry.getAttribute("filename");
+			int weight = 1;
+			try {
+				weight = Integer.parseInt(entry.getAttribute("weight"));
+			} catch (Exception e) {
+				//ignored
```

Never ever ignore exceptions, at least log them. In this context, throwing an InvalidXmlException would make sense.